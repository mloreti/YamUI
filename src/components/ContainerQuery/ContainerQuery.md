### Notes for use

You should always render a block-level element within `ContainerQuery`, this is necessary to push its width out so it can recalculate its queries on resize.


### Examples

Responsive styles/class names:

```js { "props": { "data-description": "basic" } }
const query = {
  small: { maxWidth: 500 },
  medium: { minWidth: 501, maxWidth: 700 },
  large: {minWidth: 701 },
};
const renderAt = size => {
  if (!size) {
    return <div />;
  }

  const colors = {
    small: '#A5DA51',
    medium: '#738ADD',
    large: '#DA5552',
  };

  return (
    <div className={`size-${size}`} style={{ padding: '2rem', background: colors[size], transition: 'background-color 1s', fontWeight: 'bold', color: '#fff' }}>
      This div has a different color at breakpoints below 500px (small), between 500px and 700px (medium), and above 700px (large). Currently rendering at size "{size}".
    </div>
  );
};

<div>
  <ContainerQuery query={query} renderAt={renderAt} />
</div>
```

Responsively rendering different components. This will render differently below 500px wide, between 500px and 700px wide, and wider than 700px.

```js { "props": { "data-description": "with responsive component tree" } }
const { AvatarSize } = require('../Avatar');
const { GutterSize, TextSize } = require('../Block');
const { MediaObjectSize } = require('../MediaObject');
const Like = require('../../components/Icon/icons/Like').default;
const Reply = require('../../components/Icon/icons/Reply').default;
const Share = require('../../components/Icon/icons/Share').default;
const MoreVertical = require('../../components/Icon/icons/MoreVertical').default;

const query = {
  small: { maxWidth: 500 },
  medium: { minWidth: 501, maxWidth: 700 },
  large: { minWidth: 701 },
};

const renderAt = size => {
  let avatarSize = AvatarSize.XSMALL;
  let mediaObjectSize = MediaObjectSize.XSMALL;
  let textSize = TextSize.SMALL;
  if (size === 'medium') {
    avatarSize = AvatarSize.MEDIUM;
    mediaObjectSize = MediaObjectSize.MEDIUM;
    textSize = TextSize.MEDIUM_SUB;
  } else if (size === 'large') {
    avatarSize = AvatarSize.XLARGE;
    mediaObjectSize = MediaObjectSize.XLARGE;
    textSize = TextSize.LARGE;
  }
  const avatar = <Avatar size={avatarSize} name={user.name} imageUrl={user.imageUrl} />;

  const bottomActions = size === 'small' ? null : (
    <Block textSize={TextSize.SMALL} topSpacing={GutterSize.MEDIUM}>
      <LayoutList direction="horizontal" gutterSize={GutterSize.MEDIUM}>
        <LayoutListItem>
          <Clickable><Like /> Like</Clickable>
        </LayoutListItem>
        <LayoutListItem>
          <Clickable><Reply /> Reply</Clickable>
        </LayoutListItem>
        <LayoutListItem>
          <Clickable><Share /> Share</Clickable>
        </LayoutListItem>
      </LayoutList>
    </Block>
  );

  const moreActionItems = [
    {
      key: 'like',
      text: 'Like',
      icon: Like,
      onClick: action('like'),
    },
    {
      key: 'reply',
      text: 'Reply',
      icon: Reply,
      onClick: action('reply'),
    },
    {
      key: 'share',
      text: 'Share',
      icon: Share,
      onClick: action('share'),
    },
  ];
  const moreActionsColumn = size !== 'small' ? null : (
    <FixedGridColumn fixed={true}>
      <MenuButton menuItems={moreActionItems} icon={MoreVertical} />
    </FixedGridColumn>
    
  );

  return (
    <Box>
      <FixedGridRow gutterSize={GutterSize.MEDIUM}>
        <FixedGridColumn fixed={true}>{avatar}</FixedGridColumn>
        <FixedGridColumn>
          <Block textSize={textSize}><Text bold={true}>{user.name}</Text></Block>
          <Block textSize={textSize}>hey that's a great idea! let's discuss this further at the next monthly offsite whilte the marketing team is with us.</Block>
          {bottomActions}
        </FixedGridColumn>
        {moreActionsColumn}
      </FixedGridRow>
    </Box>
  )
};

<div>
  <ContainerQuery query={query} renderAt={renderAt} />
</div>
```
