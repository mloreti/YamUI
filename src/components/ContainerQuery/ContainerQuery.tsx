/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import autobind from 'core-decorators/lib/autobind';
import { ContainerQuery as ReactContainerQuery } from 'react-container-query';
import { Props } from 'react-container-query/lib/interfaces';

export interface ContainerQueryProps {
  /**
   * An object of `size` keys mapped to height/width queries. Example:
   * {
   *    small: {
   *      maxWidth: 500,
   *    },
   *    large: {
   *      minWidth: 501,
   *    },
   * }
   */
  query: Props['query'];

  /**
   * An initial size to render at in the brief period of time while the component is first rendered
   * but the query has not yet been evaluated.
   */
  initialSize?: Props['initialSize'];

  /**
   * A render prop which receives the key of the active query.
   */
  renderAt: (size?: string) => JSX.Element;
}

interface EvaluatedQuery {
  [key: string]: boolean;
}

const getActiveQueryKey = (evaluatedQuery: EvaluatedQuery): string | undefined => {
  const keys = Object.keys(evaluatedQuery);
  // tslint:disable-next-line:no-increment-decrement
  for (let i = 0; i < keys.length; i++) {
    if (evaluatedQuery[keys[i]] === true) {
      return keys[i];
    }
  }
};

/**
 * ContainerQuery allows for conditional rendering based on the component's height or width. Provide
 * a `renderAt` prop function to render your responsive content based on the given size. For performance
 * this function will only be called when the size changes.
 */
export default class ContainerQuery extends React.Component<ContainerQueryProps> {
  public render() {
    const { query, initialSize } = this.props;
    return (
      <ReactContainerQuery query={query} initialSize={initialSize} children={this.renderSelected} />
    );
  }

  @autobind
  private renderSelected(evaluatedQuery: EvaluatedQuery) {
    return this.props.renderAt(getActiveQueryKey(evaluatedQuery));
  }
}
