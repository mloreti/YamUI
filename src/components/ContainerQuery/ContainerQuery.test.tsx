/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { mount, shallow, ReactWrapper, ShallowWrapper } from 'enzyme';
import ContainerQuery, { ContainerQueryProps } from '.';

describe('<ContainerQuery />', () => {
  let component: ShallowWrapper<ContainerQuery, {}>;
  let mounted: ReactWrapper<ContainerQuery, {}>;

  beforeEach(() => {
    const props: ContainerQueryProps = {
      query: {
        small: { maxWidth: 500 },
        large: { minWidth: 501 },
      },
      renderAt: selected => <span>{selected}</span>,
    };
    component = shallow(<ContainerQuery {...props} />);
  });

  it('matches its snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('when mounted at small size', () => {
    beforeEach(() => {
      const props: ContainerQueryProps = {
        query: {
          small: { maxWidth: 500 },
          large: { minWidth: 501 },
        },
        renderAt: selected => <div className="content">{selected}</div>,
        initialSize: { width: 400 },
      };
      mounted = mount(<ContainerQuery {...props} />);
    });

    it('matches its snapshot', () => {
      expect(mounted.find('.content').last()).toMatchSnapshot();
    });
  });

  describe('when mounted at large size', () => {
    beforeEach(() => {
      const props: ContainerQueryProps = {
        query: {
          small: { maxWidth: 500 },
          large: { minWidth: 501 },
        },
        renderAt: selected => <div className="content">{selected}</div>,
        initialSize: { width: 600 },
      };
      mounted = mount(<ContainerQuery {...props} />);
    });

    it('matches its snapshot', () => {
      expect(mounted.find('.content').last()).toMatchSnapshot();
    });
  });
});
