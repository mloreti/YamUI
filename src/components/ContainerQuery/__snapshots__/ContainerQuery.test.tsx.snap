// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContainerQuery /> matches its snapshot 1`] = `
<ContainerQuery
  query={
    Object {
      "large": Object {
        "minWidth": 501,
      },
      "small": Object {
        "maxWidth": 500,
      },
    }
  }
/>
`;

exports[`<ContainerQuery /> when mounted at large size matches its snapshot 1`] = `
<div
  className="content"
>
  large
</div>
`;

exports[`<ContainerQuery /> when mounted at small size matches its snapshot 1`] = `
<div
  className="content"
>
  small
</div>
`;
